File: leetcode/destCity.py
class Solution:
    def destCity(self, paths):
        A, B = map(set, zip(*paths))
        return list(B-A)[0]

class Solution2:
    def destCity(self, paths):
        cities = []
        for city in paths:
            cities.append(city[0])
        for city in paths:
            if city[1] not in cities:
                return(city[1])

answer = Solution2()
print(answer.destCity([["London","New York"],["New York","Lima"],["Lima","Sao Paulo"]]))
print(answer.destCity([["pYyNGfBYbm","wxAscRuzOl"],["kzwEQHfwce","pYyNGfBYbm"]]))
print(answer.destCity([["B","C"],["D","B"],["C","A"]]))
print(answer.destCity([["qMTSlfgZlC","ePvzZaqLXj"],["xKhZXfuBeC","TtnllZpKKg"],["ePvzZaqLXj","sxrvXFcqgG"],["sxrvXFcqgG","xKhZXfuBeC"],["TtnllZpKKg","OAxMijOZgW"]]))

'''
You are given the array paths, where paths[i] = [cityAi, cityBi] means there exists a direct path going from cityAi to cityBi. Return the destination city, that is, the city without any path outgoing to another city.

It is guaranteed that the graph of paths forms a line without any loop, therefore, there will be exactly one destination city.

Example 1:

Input: paths = [["London","New York"],["New York","Lima"],["Lima","Sao Paulo"]]
Output: "Sao Paulo" 
Explanation: Starting at "London" city you will reach "Sao Paulo" city which is the destination city. Your trip consist of: "London" -> "New York" -> "Lima" -> "Sao Paulo".
Example 2:

Input: paths = [["B","C"],["D","B"],["C","A"]]
Output: "A"
Explanation: All possible trips are: 
"D" -> "B" -> "C" -> "A". 
"B" -> "C" -> "A". 
"C" -> "A". 
"A". 
Clearly the destination city is "A".
Example 3:

Input: paths = [["A","Z"]]
Output: "Z"
 
Constraints:

1 <= paths.length <= 100
paths[i].length == 2
1 <= cityAi.length, cityBi.length <= 10
cityAi != cityBi
All strings consist of lowercase and uppercase English letters and the space character.


'''

File: leetcode/destCity.js
/**
 * @param {string[][]} paths
 * @return {string}
 */
var destCity = function(paths) {
  let cities = []
  for (let i = 0; i < paths.length; i++) {
    cities.push(paths[i][0])
  }
  for (let i = 0; i < paths.length; i++) {
    if (!cities.includes(paths[i][1])) {
      return paths[i][1]
    }
  }
};



console.log(destCity([["London","New York"],["New York","Lima"],["Lima","Sao Paulo"]]))
console.log(destCity([["pYyNGfBYbm","wxAscRuzOl"],["kzwEQHfwce","pYyNGfBYbm"]]))
console.log(destCity([["B","C"],["D","B"],["C","A"]]))
console.log(destCity([["qMTSlfgZlC","ePvzZaqLXj"],["xKhZXfuBeC","TtnllZpKKg"],["ePvzZaqLXj","sxrvXFcqgG"],["sxrvXFcqgG","xKhZXfuBeC"],["TtnllZpKKg","OAxMijOZgW"]]))
